### ISSUE TITLE
Support AWS SSO for S3 backend

### ISSUE THREAD

**Issue by fdw** [restic/restic#4688]
Output of `restic version`
--------------------------

`restic 0.16.2 compiled with go1.21.3 on linux/amd64`

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
Support AWS SSO logins for authenticating against the S3 backend

What are you trying to do? What problem would this solve?
---------------------------------------------------------
I would like to use restic to backup data on an AWS S3 bucket that is protected using AWS SSO. `AWS_PROFILE` is already supported by restic, but the SSO login doesn't set the env variables directly. Instead, they're written to a json file in `.aws/sso/cache`. I would like restic to read and use them (or find another way to work frictionlessly with SSO).

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Yes, restic does make me happy, so thank you for all your efforts! I feel a lot safer knowing (not just believing) that I have backups available.

**Comment by konidev20** [comment]
Hey @fdw, 

If I got your requirement correctly. 

`aws sso login --profile myprofile` creates the `.aws/sso/cache file`?

**Comment by fdw** [comment]
Yes, but it's a file in `.aws/sso/cache` with a random (as far as I can tell) name. Using the newest one should be fine.

**Comment by MichaelEischer** [comment]
We use https://github.com/minio/minio-go as library to access S3. If that has support for that SSO mechanism (not sure whether it does, I haven't checked), then I'm fine with adding support for AWS SSO.

**Comment by konidev20** [comment]
Hey found this has been discussed before on the restic forum: [Is AWS SSO authentication supported](https://forum.restic.net/t/is-aws-sso-authentication-supported/5245)

Quoting from the answer from Jul'22:
> We using the minio go sdk and not the aws sdks. From a quick glance it looks like it would be necessary to introduce a completely new authentication flow to get credentials via SSO.
>
> restic currently only supports the standard S3 access keys. If it is possible to retrieve such (temporary) keys for an SSO login, then it’s probably relatively simple to extend restic to support them. But someone would have to figure out the necessary steps and work on a pull request.

[minio-go/pkg/credentials](https://github.com/minio/minio-go/tree/master/pkg/credentials) Looks like the status remains the same. Support has not been added.

**Comment by fdw** [comment]
That is quite a pity, as I have no alternative to SSO, and copying the keys every time I want to restic is not really a viable option. It only makes me want to avoid doing backups.

Thank you anyway 🙂

**Comment by konidev20** [comment]
I have created a feature request on the minio-go github, we can probably pick this up if they're interested.

Here is a reference to that issue:
[Add ability to use AWS SSO profile as a credential provider](https://github.com/minio/minio-go/issues/1930)

**Comment by konidev20** [comment]
[aws-sso-file-credential-support.tar.gz](https://github.com/restic/restic/files/14472389/aws-sso-file-credential-support.tar.gz)

Hey @fdw, 

You can check the source code I've shared above. I was not sure how to share with your a working build, so I decided to just share the source code with you. You can run a local build.

1. Extract the tar
```
tar -xzvf aws-sso-file-credential-support.tar.gz 
```

2. Change directory into the restic folder
```
cd ./aws-sso-file-credential-support/restic
```

3. Run go mod tidy
```
go mod tidy
```

4. Run go build script
```
go run build.go
```

PS: @MichaelEischer please let me know if this is a nice way of sharing some test code, I couldn't think of anything else.

**Comment by fdw** [comment]
I just tried it, but unfortunately I only get errors: With `init`, it's `create repository at s3:s3.amazonaws.com/<bucket>/<path> failed: client.BucketExists: 301 Moved Permanently`.
With `backup`, I get:
```
create repository at s3:s3.amazonaws.com/<bucket>/<path>/ failed: client.BucketExists: 301 Moved Permanently
```

This sounds like a rights issue :/ The `aws s3` cli has no problem accessing the bucket.

**Comment by MichaelEischer** [comment]
@fdw Please use the region specific S3 endpoint, e.g. `s3.us-west-2.amazonaws.com/bucket_name`

**Comment by fdw** [comment]
Unfortunately, `s3:s3.eu-central-1.amazonaws.com/<bucket>/<path>` has the same result.

**Comment by konidev20** [comment]
@fdw I guess, you can mention this issue on the minio thread. Probably the original creator might help with something.

**Comment by jenic** [comment]
@fdw I use restic with [aws-vault](https://github.com/99designs/aws-vault) and sso (`aws-vault exec backup -- restic backup ...`), maybe that will at least allow you to use restic easily until it's officially supported?

**Comment by immortaly007** [comment]
I also ran into this issue, and worked around it using [temporary session credentials](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_use-resources.html).

This is the bash script I use now to restore the backup:
```sh
# Note that for this to work, I configured the role `backup-reader` with the appropriate Trust Relationship to allow the AWS CLI to assume the role.
AWS_CREDS=$(aws sts assume-role --role-arn arn:aws:iam::1234567890:role/backup-reader --role-session-name "ReadBackup")
AWS_ACCESS_KEY_ID=$(echo $AWS_CREDS | jq -r '.Credentials.AccessKeyId')
AWS_SECRET_ACCESS_KEY=$(echo $AWS_CREDS | jq -r '.Credentials.SecretAccessKey')
AWS_SESSION_TOKEN=$(echo $AWS_CREDS | jq -r '.Credentials.SessionToken')

RESTIC_REPOSITORY="s3:s3.eu-central-1.amazonaws.com/${S3_BACKUP_BUCKET}/"
RESTIC_PASSWORD="YourRepositoryPassword"

# You can of course also run this using your locally installed restic, in which case it's easiest to "export" the bash variables declared above so restic can see them.
docker run --rm \
  --user "$(id -u):$(id -g)" \
  -e AWS_ACCESS_KEY_ID="$AWS_ACCESS_KEY_ID" \
  -e AWS_SECRET_ACCESS_KEY="$AWS_SECRET_ACCESS_KEY" \
  -e AWS_SESSION_TOKEN="$AWS_SESSION_TOKEN" \
  -e RESTIC_REPOSITORY="$RESTIC_REPOSITORY" \
  -e RESTIC_PASSWORD="$RESTIC_PASSWORD" \
  -v "$(pwd)/restore:/target" \
  restic/restic \
  restore --no-lock latest --target /target
  ```
  
  Note that the `backup-reader` IAM role that this script uses has permissions to read my restic S3 bucket, and the following "Trust Relationship", so that the role my SSO user has, can assume the `backup-reader` role.
  ```json
  {
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Service": "s3.amazonaws.com",
                "AWS": "arn:aws:iam::1234567890:role/aws-reserved/sso.amazonaws.com/eu-central-1/AWSReservedSSO_AdministratorAccess_REDACTED"
            },
            "Action": "sts:AssumeRole"
        }
    ]
}
```

**Comment by fdw** [comment]
Seems like minio-go finally [merged support](https://github.com/minio/minio-go/issues/1930). Do you think it would be possible to update the old PRs?
